cmd/init-osm-controller/init-osm-controller.go

```diff
@@ -0,0 +1,110 @@
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const meshConfigName = "osm-mesh-config"
+
+var settings = cli.New()
+
+var (
+	flags = pflag.NewFlagSet("init-osm-controller", pflag.ExitOnError)
+	log   = logger.New("osm-controller/main")
+
+	meshName           string // An ID that uniquely identifies an OSM instance
+	kubeConfigFilePath string
+	osmNamespace       string
+
+	fetchMeshConfigMaxRetry      = 5
+	fetchMeshConfigRetryInterval = 5
+)
+
+func init() {
+	if err := logger.SetLogLevel("info"); err != nil {
+		log.Fatal().Err(err).Msg("Error setting log level")
+	}
+
+	flags.StringVar(&meshName, "mesh-name", "", "OSM mesh name")
+	flags.StringVar(&kubeConfigFilePath, "kubeconfig", "", "Path to Kubernetes config file.")
+	flags.StringVar(&osmNamespace, "osm-namespace", "", "Namespace to which OSM belongs to.")
+}
+
+func parseFlags() error {
+	if err := flags.Parse(os.Args); err != nil {
+		return err
+	}
+	_ = flag.CommandLine.Parse([]string{})
+	return nil
+}
+
+func main() {
+	log.Info().Msgf("Starting init-osm-controller %s; %s; %s", version.Version, version.GitCommit, version.BuildDate)
+
+	if err := parseFlags(); err != nil {
+		log.Fatal().Err(err).Msg("Failed parsing cmd line arguments")
+	}
+
+	// Initialize kube config and client
+	kubeConfig, err := clientcmd.BuildConfigFromFlags("", kubeConfigFilePath)
+	if err != nil {
+		log.Fatal().Err(err).Msgf("Failed creating kube config (kubeconfig=%s)", kubeConfigFilePath)
+	}
+
+	configClient, err := configClientset.NewForConfig(kubeConfig)
+	if err != nil {
+		log.Fatal().Err(err).Msgf("Could not access Kubernetes cluster, check kubeconfig.")
+		return
+	}
+
+	var retry int
+	for retry = 0; retry < fetchMeshConfigMaxRetry; retry++ {
+		if _, err := configClient.ConfigV1alpha1().MeshConfigs(settings.Namespace()).Get(context.TODO(), meshConfigName, v1.GetOptions{}); err != nil {
+			// MeshConfig does not exist. Create a default one.
+			log.Info().Msgf("MeshConfig not exist. Trying create the default MeshConfig named %s into namespace %s", meshConfigName, osmNamespace)
+
+			if err := createDefaultMeshConfig(configClient); err != nil {
+				log.Warn().Msgf("Failed creating default MeshConfig. %s. Retry after %d seconds.", err, fetchMeshConfigRetryInterval)
+
+				time.Sleep(time.Duration(fetchMeshConfigRetryInterval) * time.Second)
+			} else {
+				log.Info().Msgf("Default MeshConfig named %s created in %s.", meshConfigName, osmNamespace)
+				break
+			}
+		} else {
+			log.Info().Msgf("MeshConfig resource %s exists in %s. Skip creating MeshConfig", meshConfigName, osmNamespace)
+
+			break
+		}
+	}
+
+	if retry == fetchMeshConfigMaxRetry {
+		log.Fatal().Msgf("Max retry (%d) of fetching or creating MeshConfig reached. Exit", fetchMeshConfigMaxRetry)
+	}
+}
+
+func createDefaultMeshConfig(configClient *configClientset.Clientset) error {
+	// create meshConfig from struct
+	meshConfig := &v1alpha1.MeshConfig{}
```

Comment:
Where are the defaults for the other options being set? Currently, `enablePrometheusScraping` is set to true in `values.yaml`.

Reply:
They are defined here: https://github.com/openservicemesh/osm/blob/c0de988e82c5de275afb55876266019c38b1b461/pkg/apis/config/v1alpha1/mesh_config.go#L43

Reply:
I don't think we should be defining implementation defaults in the API definition. It seems odd to me this is the case. Wouldn't it make it easier define it in the structs at the time of default CR creation? The current encoding of the defaults in the API definition does not make this obvious.

Reply:
What if we need to create another default MeshConfig object later? Then we will need to provide the default value again.

I think, if the default values make sense to be the nature of the definition, then put it in the struct definition is fine. E.g. here, if user does not specify `prometheusScraping` value, we want it to be true rather than false.

What's your opinion?

Reply:
> What if we need to create another default MeshConfig object later? Then we will need to provide the default value again.

I don't see a use case for this.
To me, it doesn't seem correct to tie OSM install defaults to the API definition defaults. There is a more subtle reason for this. At some point when our API is mature and there is a change to the API, the API version will need to be updated (ex. from v1alpha1 to v1alpha2). So if we decide to update the install time defaults, then the API version will also require to be updated which is not desirable. This seems to not be a problem because users are not using the MeshConfig CR, but once this is released it will be better to control install time defaults via the application code rather than encoding it in the API definition.

Reply:
Ok, make sense to me

Reply:
FYI, what I mentioned was not accurate. The code `meshConfig` was later overwritten by k8s api creation. Thus the default value is from the CRD definition.

Reply:
I didn't get the last comment. What is the final decision?